Update and improve BBCode in RichTextlabel documentation for Godot 4.0 (#6991)
godotengine/godot-docs: tutorials/ui/bbcode_in_richtextlabel.rst

:ref:`class_Label` nodes are great for displaying basic text, but they have limitations.
If you want to change the color of the text, or its alignment, you can only do that to
the entire label. You can't make a part of the text have another color, or have a part
of the text centered. To get around these limitations, you would use a :ref:`class_RichTextLabel`.

:ref:`class_RichTextLabel` allows for complex formatting of text using a markup syntax or
the built-in API. It uses BBCodes for the markup syntax, a system of tags that designate
formatting rules for a part of the text. You may be familiar with them if you ever used
forums (also known as `bulletin boards`, hence the "BB" in "BBCode").

Note that the BBCode tags can also be used to some extent in the XML source of
the class reference. For more information, see :ref:`doc_class_reference_primer`.

.. seealso::

    You can see how BBCode in RichTextLabel works in action using the
    `Rich Text Label with BBCode demo project <https://github.com/godotengine/godot-demo-projects/tree/master/gui/rich_text_bbcode>`__.

Using BBCode
------------

> By default, :ref:`class_RichTextLabel` functions exactly the same as a normal :ref:`class_Label`.

Reviewer: ```suggestion
By default, :ref:`class_RichTextLabel` functions like a normal :ref:`class_Label`.
```
There are some differences between default `RichTextLabel` and `Label`, such as the minimum size and scrollbar.

the content can be omitted.

.. code-block:: none

  [tag]content[/tag]
  [tag=value]content[/tag]
  [tag option1=value1 option2=value2]content[/tag]
  [tag][/tag]
  [tag]

.. note::

    RichTextLabel doesn't support entangled BBCode tags. For example, instead of
    using:

    ::

        [b]bold[i]bold italic[/b]italic[/i]

    Use:

    ::

        [b]bold[i]bold italic[/i][/b][i]italic[/i]

.. _doc_bbcode_in_richtextlabel_handling_user_input_safely:

Handling user input safely
--------------------------

In a scenario where users may freely input text (such as chat in a multiplayer
game), you should make sure users cannot use arbitrary BBCode tags that will be
parsed by RichTextLabel. This is to avoid inappropriate use of formatting, which
can be problematic if ``[url]`` tags are handled by your RichTextLabel (as players
may be able to create clickable links to phishing sites or similar).

Using RichTextLabel's ``[lb]`` and/or ``[rb]`` tags, we can replace the opening and/or
closing brackets of any BBCode tag in a message with those escaped tags. This
prevents users from using BBCode that will be parsed as tags – instead, the

> BBCode will be displayed as text.

Reviewer: Perhaps it's worth mentioning the alternative to _remove_ tags with a regular expression?

```gdscript
var regex = RegEx.new()
regex.compile("\\[.*?\\]")
text = regex.sub(text, "", true)
```

Author: I'm not sure. In some cases, stripping tags is desired, but for something like multiplayer chat, it can have unintended effects. For example, if I write a message `[IMPORTANT] Hello world`, I don't expect the `[IMPORTANT]` to disappear entirely.
